vita3k/host/src/pkg.cpp

```diff
@@ -0,0 +1,346 @@
+// Vita3K emulator project
+// Copyright (C) 2020 Vita3K team
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+#include <PsvPfsParserConfig.h>
+#include <app/functions.h>
+#include <crypto/aes.h>
+#include <host/functions.h>
+#include <host/pkg.h>
+#include <host/sce_types.h>
+#include <host/sfo.h>
+#include <host/state.h>
+#include <io/device.h>
+#include <rif2zrif.h>
+#include <util/bytes.h>
+#include <util/log.h>
+
+// Credits to mmozeiko https://github.com/mmozeiko/pkg2zip
+
+static void ctr_add(uint8_t *counter, uint64_t n) {
+    for (int i = 15; i >= 0; i--) {
+        n = n + counter[i];
+        counter[i] = (uint8_t)n;
+        n >>= 8;
+    }
+}
+
+static void aes128_ctr_xor(aes_context *ctx, const uint8_t *iv, uint64_t block, uint8_t *input, size_t size) {
+    uint8_t tmp[16];
+    uint8_t counter[16];
+    for (uint32_t i = 0; i < 16; i++) {
+        counter[i] = iv[i];
+    }
+    ctr_add(counter, block);
+
+    while (size >= 16) {
+        aes_crypt_ecb(ctx, AES_ENCRYPT, counter, tmp);
+        for (uint32_t i = 0; i < 16; i++) {
+            *input++ ^= tmp[i];
+        }
+        ctr_add(counter, 1);
+        size -= 16;
+    }
+
+    if (size != 0) {
+        aes_crypt_ecb(ctx, AES_ENCRYPT, counter, tmp);
+        for (size_t i = 0; i < size; i++) {
+            *input++ ^= tmp[i];
+        }
+    }
+}
+
+bool decrypt_install_nonpdrm(std::string &drmlicpath, const std::string &title_path, HostState &host) {
+    host.zRIF = rif2zrif(drmlicpath);
+    host.title_id_src = title_path;
+    host.title_id_dst = title_path + "_dec";
+    host.f00d_enc_type = F00DEncryptorTypes::native;
+    host.f00d_arg = std::string();
```

Comment:
If these are cleared at the end of the function why can't these be local variables? :)

Reply:
Uhhh host.zrif is captured by the gui.

Reply:
it would be best to return stuff to the gui, or even pass GuiState here if you are going to display it- especially if this stuff is really only meant to be displayed in the gui (host is a god struct used in every sce module, will these variables be used anywhere other than gui?)